From 3e4a94a92d604c16102c8d0daee2d17155575050 Mon Sep 17 00:00:00 2001
From: W-Mark Kubacki <wmark@hurrikane.de>
Date: Sat, 10 Jan 2015 15:11:39 +0100
Subject: [PATCH] Set content-type header with charset for .html files.

This enables correct rendering of such files containing UTF-8 runes.
---
 src/HttpServer.cc | 7 ++++++-
 src/HttpServer.h  | 1 +
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/HttpServer.cc b/src/HttpServer.cc
index b9b404d..b5e495c 100644
--- a/src/HttpServer.cc
+++ b/src/HttpServer.cc
@@ -673,6 +673,7 @@ int Http2Handler::verify_npn_result() {
 int Http2Handler::submit_file_response(const std::string &status,
                                        Stream *stream, time_t last_modified,
                                        off_t file_length,
+                                       const std::string path,
                                        nghttp2_data_provider *data_prd) {
   std::string content_length = util::utos(file_length);
   std::string last_modified_str;
@@ -683,12 +684,16 @@ int Http2Handler::submit_file_response(const std::string &status,
       http2::make_nv_ll("cache-control", "max-age=3600"),
       http2::make_nv_ls("date", sessions_->get_cached_date()),
       http2::make_nv_ll("", ""),
+      http2::make_nv_ll("", ""),
   };
   size_t nvlen = 5;
   if (last_modified != 0) {
     last_modified_str = util::http_date(last_modified);
     nva[nvlen++] = http2::make_nv_ls("last-modified", last_modified_str);
   }
+  if (nghttp2::util::endsWith(path, ".html")) {
+    nva[nvlen++] = http2::make_nv_ll("content-type", "text/html; charset=UTF-8");
+  }
   return nghttp2_submit_response(session_, stream->stream_id, nva, nvlen,
                                  data_prd);
 }
@@ -999,7 +1004,7 @@ void prepare_response(Stream *stream, Http2Handler *hd,
   }
 
   hd->submit_file_response(STATUS_200, stream, buf.st_mtime, buf.st_size,
-                           &data_prd);
+                           path, &data_prd);
 }
 } // namespace
 
diff --git a/src/HttpServer.h b/src/HttpServer.h
index 87d3296..8f88e7b 100644
--- a/src/HttpServer.h
+++ b/src/HttpServer.h
@@ -104,6 +104,7 @@ public:
 
   int submit_file_response(const std::string &status, Stream *stream,
                            time_t last_modified, off_t file_length,
+                           const std::string path,
                            nghttp2_data_provider *data_prd);
 
   int submit_response(const std::string &status, int32_t stream_id,
-- 
2.0.4

